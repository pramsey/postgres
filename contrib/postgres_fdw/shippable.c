/*-------------------------------------------------------------------------
 *
 * shippable.c
 *	  Non-built-in objects cache management and utilities.
 *
 * Is a non-built-in shippable to the remote server? Only if
 * the object is in an extension declared by the user in the
 * OPTIONS of the wrapper or the server.
 *
 * Portions Copyright (c) 1996-2015, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *	  contrib/postgres_fdw/shippable.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include "postgres_fdw.h"

#include "access/genam.h"
#include "access/heapam.h"
#include "access/htup_details.h"
#include "catalog/dependency.h"
#include "catalog/indexing.h"
#include "catalog/pg_depend.h"
#include "commands/extension.h"
#include "utils/builtins.h"
#include "utils/fmgroids.h"
#include "utils/hsearch.h"
#include "utils/rel.h"
#include "utils/snapmgr.h"

/* Hash table for informations about remote objects we'll call */
static HTAB *ShippableCacheHash = NULL;

/* objid is the lookup key, must appear first */
typedef struct
{
	Oid	objid;
} ShippableCacheKey;

typedef struct
{
	ShippableCacheKey key; /* lookup key - must be first */
	bool shippable;        /* extension the object appears within, or InvalidOid if none */
} ShippableCacheEntry;

/*
 * InitializeShippableCache
 *	    Initialize the cache of functions we can ship to remote server.
 */
static void
InitializeShippableCache(void)
{
	HASHCTL ctl;

	/* Initialize the hash table. */
	MemSet(&ctl, 0, sizeof(ctl));
	ctl.keysize = sizeof(ShippableCacheKey);
	ctl.entrysize = sizeof(ShippableCacheEntry);
	ShippableCacheHash =
		hash_create("Shippable cache", 256, &ctl, HASH_ELEM);
}

/*
 * Returns true if given operator/function is part of an extension declared in the
 * server options.
 */
static bool
lookup_shippable(Oid objnumber, List *extension_list)
{
	static int nkeys = 1;
	ScanKeyData key[nkeys];
	HeapTuple tup;
	Relation depRel;
	SysScanDesc scan;
	int nresults = 0;

	/* Always return false if we don't have any declared extensions */
	if (!extension_list)
		return false;

	/* We need this relation to scan */
	depRel = heap_open(DependRelationId, RowExclusiveLock);

	/* Scan the system dependency table for all entries this object */
	/* depends on, then iterate through and see if one of them */
	/* is an extension declared by the user in the options */
	ScanKeyInit(&key[0],
				Anum_pg_depend_objid,
				BTEqualStrategyNumber, F_OIDEQ,
				ObjectIdGetDatum(objnumber));

	scan = systable_beginscan(depRel, DependDependerIndexId, true,
							  GetCatalogSnapshot(depRel->rd_id), nkeys, key);

	while (HeapTupleIsValid(tup = systable_getnext(scan)))
	{
		Form_pg_depend foundDep = (Form_pg_depend) GETSTRUCT(tup);

		if (foundDep->deptype == DEPENDENCY_EXTENSION)
		{
			ListCell *ext;

			foreach(ext, extension_list)
			{
				Oid extension_oid = (Oid) lfirst(ext);
				if (foundDep->refobjid == extension_oid)
				{
					nresults++;
				}
			}
		}
		if (nresults > 0) break;
	}

	systable_endscan(scan);
	relation_close(depRel, RowExclusiveLock);

	return nresults > 0;
}

/*
 * is_shippable
 *     Is this object (proc/op/type) shippable to foreign server?
 *     Check cache first, then look-up whether (proc/op/type) is
 *     part of a declared extension if it is not cached.
 */
bool
is_shippable(Oid objnumber, PgFdwRelationInfo *fpinfo)
{
	ShippableCacheKey key;
	ShippableCacheEntry *entry;

	/* Always return false if we don't have any declared extensions */
	if (!fpinfo->extensions)
		return false;

	/* Find existing cache, if any. */
	if (!ShippableCacheHash)
		InitializeShippableCache();

	/* Zero out the key */
	memset(&key, 0, sizeof(key));

	key.objid = objnumber;

	entry = (ShippableCacheEntry *)
				 hash_search(ShippableCacheHash,
					(void *) &key,
					HASH_FIND,
					NULL);

	/* Not found in ShippableCacheHash cache.  Construct new entry. */
	if (!entry)
	{
		/* Right now "shippability" is exclusively a function of whether */
		/* the obj (proc/op/type) is in an extension declared by the user. In the future */
		/* we could additionally have a whitelist of functions declared one */
		/* at a time. */
		bool shippable = lookup_shippable(objnumber, fpinfo->extensions);

		entry = (ShippableCacheEntry *)
					 hash_search(ShippableCacheHash,
						(void *) &key,
						HASH_ENTER,
						NULL);

		entry->shippable = shippable;
	}

	if (!entry)
		return false;
	else
		return entry->shippable;
}

/*
 * extractExtensionList
 *     Parse a comma-separated string and fill out the list
 *     argument with the Oids of the extensions in the string.
 *     If an extenstion provided cannot be looked up in the
 *     catalog (it hasn't been installed or doesn't exist)
 *     then throw an error.
 */
bool
extractExtensionList(char *extensionString, List **extensionOids)
{
	List *extlist;
	ListCell   *l, *o;

	if (!SplitIdentifierString(extensionString, ',', &extlist))
	{
		list_free(extlist);
		ereport(ERROR,
			(errcode(ERRCODE_SYNTAX_ERROR),
			 errmsg("unable to parse extension list \"%s\"",
				extensionString)));
	}

	foreach(l, extlist)
	{
		const char *extension_name = (const char *) lfirst(l);
		Oid extension_oid = get_extension_oid(extension_name, true);
		if (extension_oid == InvalidOid)
		{
			ereport(ERROR,
				(errcode(ERRCODE_SYNTAX_ERROR),
				 errmsg("the \"%s\" extension must be installed locally before it can be used on a remote server",
					extension_name)));
		}
		/* Option validation calls this function with NULL in the */
		/* extensionOids parameter, to just do existence/syntax */
		/* checking of the option */
		else if (extensionOids)
		{
			bool found = false;
			/* Only add this extension Oid to the list */
			/* if we don't already have it in the list */
			foreach(o, *extensionOids)
			{
				Oid oid = (Oid) lfirst(o);
				if (oid == extension_oid)
					found = true;
			}
			if (!found)
				*extensionOids = lappend_oid(*extensionOids, extension_oid);
		}
	}

	list_free(extlist);
	return true;
}
