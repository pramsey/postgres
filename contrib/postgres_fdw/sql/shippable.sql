-- ===================================================================
-- create FDW objects
-- ===================================================================

-- Error, extension isn't installed yet
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');

-- Try again
CREATE EXTENSION cube;
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
ALTER SERVER loopback OPTIONS (DROP extensions);


-- ===================================================================
-- create objects used through FDW loopback server
-- ===================================================================

CREATE SCHEMA "SH 1";
CREATE TABLE "SH 1"."TBL 1" (
	"C 1" int NOT NULL,
	c2 int NOT NULL,
	c3 cube,
	c4 timestamptz
);

INSERT INTO "SH 1"."TBL 1"
	SELECT id,
	       2 * id,
	       cube(id,2*id),
	       '1970-01-01'::timestamptz + ((id % 100) || ' days')::interval
	FROM generate_series(1, 1000) id;

ANALYZE "SH 1"."TBL 1";

-- ===================================================================
-- create foreign table
-- ===================================================================

CREATE FOREIGN TABLE shft1 (
	"C 1" int NOT NULL,
	c2 int NOT NULL,
	c3 cube,
	c4 timestamptz
) SERVER loopback
OPTIONS (schema_name 'SH 1', table_name 'TBL 1');

-- ===================================================================
-- simple queries
-- ===================================================================

-- without operator shipping
EXPLAIN (COSTS false) SELECT * FROM shft1 LIMIT 1;
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;

-- with operator shipping
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
EXPLAIN VERBOSE SELECT cube_dim(c3) FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
SELECT cube_dim(c3) FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE cube_dim(c3) = 1 LIMIT 2;
SELECT c2 FROM shft1 WHERE cube_dim(c3) = 1 LIMIT 2;



