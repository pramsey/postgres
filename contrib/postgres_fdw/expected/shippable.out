-- ===================================================================
-- create FDW objects
-- ===================================================================
-- Error, extension isn't installed yet
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
ERROR:  the "cube" extension must be installed locally before it can be used on a remote server
-- Try again
CREATE EXTENSION cube;
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
ALTER SERVER loopback OPTIONS (DROP extensions);
-- ===================================================================
-- create objects used through FDW loopback server
-- ===================================================================
CREATE SCHEMA "SH 1";
CREATE TABLE "SH 1"."TBL 1" (
	"C 1" int NOT NULL,
	c2 int NOT NULL,
	c3 cube,
	c4 timestamptz
);
INSERT INTO "SH 1"."TBL 1"
	SELECT id,
	       2 * id,
	       cube(id,2*id),
	       '1970-01-01'::timestamptz + ((id % 100) || ' days')::interval
	FROM generate_series(1, 1000) id;
ANALYZE "SH 1"."TBL 1";
-- ===================================================================
-- create foreign table
-- ===================================================================
CREATE FOREIGN TABLE shft1 (
	"C 1" int NOT NULL,
	c2 int NOT NULL,
	c3 cube,
	c4 timestamptz
) SERVER loopback
OPTIONS (schema_name 'SH 1', table_name 'TBL 1');
-- ===================================================================
-- simple queries
-- ===================================================================
-- without operator shipping
EXPLAIN (COSTS false) SELECT * FROM shft1 LIMIT 1;
         QUERY PLAN          
-----------------------------
 Limit
   ->  Foreign Scan on shft1
(2 rows)

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
                             QUERY PLAN                              
---------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..205.06 rows=15 width=4)
   Output: c2
   Filter: (shft1.c3 && '(1.5),(2.5)'::cube)
   Remote SQL: SELECT c2, c3 FROM "SH 1"."TBL 1"
(4 rows)

SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
 c2 
----
  2
  4
(2 rows)

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..205.06 rows=15 width=4)
   Output: c2
   Filter: (shft1.c3 && '(1.5),(2.5)'::cube)
   Remote SQL: SELECT c2, c3 FROM "SH 1"."TBL 1"
(4 rows)

-- with operator shipping
ALTER SERVER loopback OPTIONS (ADD extensions 'cube');
EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..146.86 rows=15 width=4)
   Output: c2
   Remote SQL: SELECT c2 FROM "SH 1"."TBL 1" WHERE ((c3 OPERATOR(public.&&) '(1.5),(2.5)'::public.cube))
(3 rows)

SELECT c2 FROM shft1 WHERE c3 && cube(1.5,2.5);
 c2 
----
  2
  4
(2 rows)

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..146.86 rows=15 width=4)
   Output: c2
   Remote SQL: SELECT c2 FROM "SH 1"."TBL 1" WHERE ((c3 OPERATOR(public.&&) '(1.5),(2.5)'::public.cube))
(3 rows)

EXPLAIN VERBOSE SELECT cube_dim(c3) FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Foreign Scan on public.shft1  (cost=100.00..128.43 rows=7 width=32)
   Output: cube_dim(c3)
   Remote SQL: SELECT c3 FROM "SH 1"."TBL 1" WHERE ((c3 OPERATOR(public.&&) '(1.5),(2.5)'::public.cube))
(3 rows)

SELECT cube_dim(c3) FROM shft1 WHERE c3 && '(1.5),(2.5)'::cube;
 cube_dim 
----------
        1
        1
(2 rows)

EXPLAIN VERBOSE SELECT c2 FROM shft1 WHERE cube_dim(c3) = 1 LIMIT 2;
                                     QUERY PLAN                                      
-------------------------------------------------------------------------------------
 Limit  (cost=100.00..107.22 rows=2 width=4)
   Output: c2
   ->  Foreign Scan on public.shft1  (cost=100.00..154.18 rows=15 width=4)
         Output: c2
         Remote SQL: SELECT c2 FROM "SH 1"."TBL 1" WHERE ((public.cube_dim(c3) = 1))
(5 rows)

SELECT c2 FROM shft1 WHERE cube_dim(c3) = 1 LIMIT 2;
 c2 
----
  2
  4
(2 rows)

-- ===================================================================
-- clean up
-- ===================================================================
